File: elixir/lib/day01.ex
defmodule AdventOfCode2023.Day01 do
  @moduledoc """
  Day 1: Trebuchet?!

  https://adventofcode.com/2023/day/1
  """

  def main do
    input =
      "../data/01-input.txt"
      |> File.read!()
      |> String.split("\n", trim: true)

    IO.inspect({
      part_one(input),
      part_two(input)
    })
  end

  def part_one(lines) do
    lines
    |> Enum.reduce([], fn line, acc ->
      matches =
        line
        |> String.split("", trim: true)
        |> Enum.with_index()
        |> Enum.reduce(%{}, fn {char, index}, acc ->
          if Regex.match?(~r/\d/, char) do
            # char is a digit, put it in the map
            Map.put(acc, index, char)
          else
            acc
          end
        end)

      {min, max} =
        {matches
         |> Map.keys()
         |> Enum.min(),
         matches
         |> Map.keys()
         |> Enum.max()}

      calibration = String.to_integer(matches[min] <> matches[max])
      [calibration | acc]
    end)
    |> Enum.sum()
  end

  def part_two(lines) do
    lookup = %{
      "one" => "1",
      "two" => "2",
      "three" => "3",
      "four" => "4",
      "five" => "5",
      "six" => "6",
      "seven" => "7",
      "eight" => "8",
      "nine" => "9"
    }

    calibrations =
      Enum.reduce(lines, [], fn line, acc ->
        # build a map of matching indices to numbers, e.g.
        # "two1nine" => %{0 => "2", 3 => "1", 4 => "9"}
        matches =
          line
          |> String.split("", trim: true)
          |> Enum.with_index()
          |> Enum.reduce(%{}, fn {char, index}, acc ->
            if Regex.match?(~r/\d/, char) do
              # char is a digit, put it in the map
              Map.put(acc, index, char)
            else
              slice = String.slice(line, index..-1)

              Enum.reduce_while(lookup, acc, fn {word, number}, acc ->
                if String.starts_with?(slice, word) do
                  # we're at a numeric word, put number in the map
                  {:halt, Map.put(acc, index, number)}
                else
                  {:cont, acc}
                end
              end)
            end
          end)

        # take min/max indexes from matches map as first/last index
        {first_idx, last_idx} =
          {matches
           |> Map.keys()
           |> Enum.min(),
           matches
           |> Map.keys()
           |> Enum.max()}

        # concat first/last index vals and convert to integer for calibration
        calibration = String.to_integer(matches[min_idx] <> matches[max_idx])

        [calibration | acc]
      end)

    Enum.sum(calibrations)
  end
end

File: rust/src/bin/day01.rs
use advent_of_code_2023::*;
use std::collections::HashMap;

// Day 1: Trebuchet?!
// https://adventofcode.com/2023/day/1

fn main() {
    // Part One
    // ========
    let lines = read_lines_from_file("01-input.txt");

    let mut calibrations: Vec<i32> = Vec::new();

    lines.iter().for_each(|line| {
        let mut first: Option<char> = None;
        let mut last: Option<char> = None;

        for c in line.chars() {
            if c.is_numeric() {
                if first.is_none() {
                    first = Some(c);
                } else {
                    last = Some(c);
                }
            }
        }

        if last.is_none() {
            last = first;
        }

        // horrifying
        let calibration = (first.unwrap().to_string() + &last.unwrap().to_string())
            .parse::<i32>()
            .unwrap();

        calibrations.push(calibration);
    });

    let sum: i32 = calibrations.iter().sum();
    println!("{:?}", sum);

    // Part Two
    // ========
    let lines = read_lines_from_file("01-input.txt");

    let nums = ["one", "two", "three", "four", "five", "six", "seven", "eight", "nine"];
    let mut calibrations: Vec<i32> = Vec::new();

    for line in &lines {
        // build a map of matching indexes to numbers
        // e.g. "two1nine" would be:
        //   {
        //     0: 2,
        //     3: 1,
        //     4: 9
        //   }

        let mut matches: HashMap<usize, usize> = HashMap::new();
        let mut skip_until = None;

        for (i, char) in line.char_indices() {
            // optimization skip
            if let Some(skip_index) = skip_until {
                if i < skip_index {
                    continue;
                }
                skip_until = None;
            }

            // if the character is numeric, put that as match and continue
            if char.is_numeric() {
                let num = char.to_digit(10).unwrap() as usize;
                matches.insert(i, num);
                continue;
            } else {
                // else loop through nums and see if any word matches
                for (j, word) in nums.iter().enumerate() {
                    let num = j + 1;
                    if line[i..].starts_with(word) {
                        matches.insert(i, num);
                        skip_until = Some(i + word.len() - 1);
                        break;
                    }
                }
            }
        }

        // get values in matches map for min/max keys
        // e.g. "two1nine" matches => {0: 2, 3: 1, 4: 9}
        // min key is 0 with a 2 value, max key is 4, with a 9 value
        let first = matches
            .keys()
            .min()
            .and_then(|&min_key| matches.get(&min_key))
            .unwrap();

        let last = matches
            .keys()
            .max()
            .and_then(|&max_key| matches.get(&max_key))
            .unwrap();

        // concat first and last as strings and convert to i32 for calibration
        let calibration = format!("{}{}", first, last).parse::<i32>().unwrap();

        calibrations.push(calibration);
    }

    let sum: i32 = calibrations.iter().sum();
    println!("{:?}", sum);
}
